#! /bin/sh /usr/share/dpatch/dpatch-run
## 65-sieveshell-enhancements.dpatch by Sven Mueller <debian@incase.de>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Adds some enhancements to sieveshell and fixes some paths.

@DPATCH@
--- a/perl/imap/IMAP/Shell.pm
+++ b/perl/imap/IMAP/Shell.pm
@@ -126,7 +126,7 @@ my %builtins = (exit =>
 		  [\&_sc_info, '[mailbox]',
 		   'display mailbox/server metadata'],
 		mboxcfg =>
-		  [\&_sc_mboxcfg, 'mailbox [comment|news2mail|expire|sieve|squat] value',
+		  [\&_sc_mboxcfg, 'mailbox [comment|news2mail|expire|squat|/<explicit annotation>] value',
 		   'configure mailbox'],
 		mboxconfig => 'mboxcfg',
 		reconstruct =>
@@ -437,7 +437,7 @@ sub run {
 # programs, as opposed to things expected from within a program.)
 sub shell {
   my ($server, $port, $authz, $auth, $systemrc, $userrc, $dorc, $mech, $pw) =
-    ('', 143, undef, $ENV{USER} || $ENV{LOGNAME}, '/usr/local/etc/cyradmrc.pl',
+    ('', 143, undef, $ENV{USER} || $ENV{LOGNAME}, '/etc/cyradmrc.pl',
      "$ENV{HOME}/.cyradmrc.pl", 1, undef, undef);
   GetOptions('user|u=s' => \$auth,
 	     'authz|z=s' => \$authz,
@@ -467,7 +467,7 @@ sub shell {
 			  -rock => \$cyradm});
     $cyradm->authenticate(-authz => $authz, -user => $auth,
 			  -mechanism => $mech, -password => $pw)
-      or die "cyradm: cannot authenticate to server with $mech as $auth\n";
+      or die "cyradm: cannot authenticate to server" . (defined($mech)?" with $mech":"") . " as $auth\n";
   }
   my $fstk = [*STDIN, *STDOUT, *STDERR];
   if ($dorc && $systemrc ne '' && -f $systemrc) {
@@ -1400,7 +1400,7 @@ sub _sc_mboxcfg {
   while (defined ($opt = shift(@argv))) {
     last if $opt eq '--';
     if ($opt =~ /^-/) {
-      die "usage: mboxconfig mailbox [comment|news2mail|expire|sieve|squat] value\n";
+      die "usage: mboxconfig mailbox [comment|news2mail|expire|squat|/<explicit annotation>] value\n";
     }
     else {
       push(@nargv, $opt);
@@ -1409,7 +1409,7 @@ sub _sc_mboxcfg {
   }
   push(@nargv, @argv);
   if (@nargv < 2) {
-    die "usage: mboxconfig mailbox [comment|news2mail|expire|sieve|squat] value\n";
+    die "usage: mboxconfig mailbox [comment|news2mail|expire|squat|/<explicit annotation>] value\n";
   }
   if (!$cyrref || !$$cyrref) {
     die "mboxconfig: no connection to server\n";
--- a/perl/sieve/lib/request.c
+++ b/perl/sieve/lib/request.c
@@ -560,7 +560,6 @@ int getscript(int version, struct protst
 	      char **refer_to, char **errstrp)
 {
   int res;
-  mystring_t *str=NULL;
   mystring_t *errstr=NULL;
   lexstate_t state;
   int ret = 0;
--- a/perl/sieve/scripts/sieveshell.pl
+++ b/perl/sieve/scripts/sieveshell.pl
@@ -60,7 +60,9 @@ my $deletehelp =     "delete <name>    -
 my $username = $ENV{USER};
 my $authname = $ENV{USER};
 my $realm = "";
+my $password;
 my $ex = "";
+my $exfile = "";
 my $help = 0;
 my $man = 0;
 my $ret;
@@ -68,7 +70,9 @@ my $ret;
 GetOptions("a|authname:s" => \$authname,
     "u|username:s" => \$username,
     "r|realm:s" => \$realm,
+    "p|password:s" => \$password,
     "e|exec:s" => \$ex,
+    "f|execfile:s" => \$exfile,
     "help|?" => \$help,
     man => \$man) or pod2usage(2);
 pod2usage(1) if $help;
@@ -83,20 +87,24 @@ my $acapserver = $ARGV[0];
 my $filehandle;
 my $interactive;
 
-if (! $ex eq "") {
-    $filehandle = tempfile();
-
-    if (!$filehandle) { die "unable to open tmp file: $?"; }
-
-    print $filehandle $ex;
-    seek $filehandle, 0, 0; # rewind file
+if (! $exfile eq "") {
+    open(FILEH,"<$exfile") || die "unable to open file: $?";
+    $filehandle = *FILEH;
     $interactive = 0;
 } else {
-    $filehandle = *STDIN;
-    $interactive = 1;
-}
+    if (! $ex eq "") {
+	$filehandle = tempfile();
 
+	if (!$filehandle) { die "unable to open tmp file: $?"; }
 
+	print $filehandle $ex;
+	seek $filehandle, 0, 0; # rewind file
+	$interactive = 0;
+    } else {
+	$filehandle = *STDIN;
+	$interactive = 1;
+    }
+}
 
 sub list_cb {
 
@@ -121,6 +129,8 @@ sub prompt {
       return $authname;
   } elsif (($type eq "realm") && (defined $realm)) {
       return $realm;
+  } elsif (($type eq "password") && (defined $password)) {
+      return $password;
   }
 
   my $ostty;
@@ -171,6 +181,8 @@ if (!defined $obj) {
 
 my $term = Term::ReadLine->new("sieveshell");
 
+my $exitcode = 0;
+
 $term->ornaments(0);
 
 while(defined($_  = ($interactive ? $term->readline('> ') : <$filehandle>))){
@@ -197,6 +209,9 @@ while(defined($_  = ($interactive ? $ter
 	my $errstr = sieve_get_error($obj);
 	$errstr = "unknown error" if(!defined($errstr));
 	print "upload failed: $errstr\n"; 
+	$exitcode = 1;
+      } else {
+        $exitcode = 0;
       }
     } elsif (($words[0] eq "list") || 
 	     ($words[0] eq "l") || 
@@ -206,6 +221,9 @@ while(defined($_  = ($interactive ? $ter
 	    my $errstr = sieve_get_error($obj);
 	    $errstr = "unknown error" if(!defined($errstr));
 	    print "list failed: $errstr\n";
+	    $exitcode = 1;
+	} else {
+	    $exitcode = 0;
 	}
     } elsif (($words[0] eq "activate") || 
 	     ($words[0] eq "a")) {
@@ -218,6 +236,9 @@ while(defined($_  = ($interactive ? $ter
 	    my $errstr = sieve_get_error($obj);
 	    $errstr = "unknown error" if(!defined($errstr));
 	    print "activate failed: $errstr\n";
+	    $exitcode = 1;
+	} else {
+	    $exitcode = 0;
 	}
     } elsif (($words[0] eq "deactivate") || 
 	     ($words[0] eq "da")) {
@@ -230,6 +251,9 @@ while(defined($_  = ($interactive ? $ter
 	    my $errstr = sieve_get_error($obj);
 	    $errstr = "unknown error" if(!defined($errstr));
 	    print "deactivate failed: $errstr\n";
+	    $exitcode = 1;
+	} else {
+	    $exitcode = 0;
 	}
     } elsif (($words[0] eq "delete") || 
 	     ($words[0] eq "d")) {    
@@ -242,6 +266,9 @@ while(defined($_  = ($interactive ? $ter
 	    my $errstr = sieve_get_error($obj);
 	    $errstr = "unknown error" if(!defined($errstr));
 	    print "delete failed: $errstr\n"; 
+	    $exitcode = 1;
+	} else {
+	    $exitcode = 0;
 	}
     } elsif (($words[0] eq "get") || 
 	     ($words[0] eq "g")) {
@@ -255,25 +282,32 @@ while(defined($_  = ($interactive ? $ter
 	    my $errstr = sieve_get_error($obj);
 	    $errstr = "unknown error" if(!defined($errstr));
 	    print "get failed: $errstr\n"; 
+	    $exitcode = 1;
 	} else {
 	    if ($words[2]) {
 		open (OUTPUT,">$words[2]") || die "Unable to open $words[2]";
 		print OUTPUT $str;
 		close(OUTPUT);
+		$exitcode = 0;
 	    } else {
 		print $str;
+		$exitcode = 0;
 	    }
 	}
     } elsif (($words[0] eq "quit") || ($words[0] eq "q")) {
         sieve_logout($obj);
-	exit 0;
+	exit $exitcode;
     } elsif (($words[0] eq "help") || ($words[0] eq "?")) {
 	show_help();
+	$exitcode = 0;
     } else {
 	print "Invalid command: $words[0]\n";
+	$exitcode = 1;
     } 
 }
 
+exit $exitcode;
+
 __END__
 
 =head1 NAME
@@ -283,7 +317,8 @@ sieveshell - remotely manipulate sieve s
 =head1 SYNOPSIS
 
 sieveshell [B<--user>=I<user>] [B<--authname>=I<authname>] 
-[B<--realm>=I<realm>] [B<--exec>=I<script>] I<server>[B<:>I<port>]
+[B<--realm>=I<realm>] [B<--password>=I<password>]
+[B<--exec>=I<script>] [B<--execfile>=I<file>] I<server>[B<:>I<port>]
 
 sieveshell B<--help>
 
@@ -327,11 +362,21 @@ The user to use for authentication (defa
 
 The realm to attempt authentication in.
 
+=item B<-p> I<password>, B<--password>=I<password>
+
+The password to use when authenticating to server. Note that this
+parameter can be seen in the process list. B<Use with caution!>
+
 =item B<-e> I<script>, B<--exec>=I<script> 
 
 Instead of working interactively, run commands from I<script>, and
 exit when done.
 
+=item B<-f> I<file>, B<--execfile>=I<file>
+
+Instead of working interactively, run commands from file I<file> and
+exit when done.
+
 =back
 
 =head1 REFERENCES
